### PR TITLE
[WIP]FunctionalEncryptedTensor Shell

### LINKED ISSUE #3581: Create the FunctionalEncrypted Tensor shell

### PR DESCRIPTION
**Pull request by sukhadj**
# Pull Request

## Description
Fixes #3581 

This PR creates the initial shell required for FunctionalEncryptedTensor.
Allows encryption of the tensor.  
Also added an experimental notebook for functional encryption.

- [x] Add pyfe as a dependency 
- [x] Add tests

## Affected Dependencies
Related PR : https://github.com/OpenMined/PyFE/pull/1

## Type of Change
Please mark options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (non-breaking change which adds documentation)
- [ ] Improvement (non-breaking change that improves the performance or reliability of existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
Not tested yet

## Checklist
- [x] I did follow the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md)
- [x] I did follow the [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have added tests for my changes

## Additional Context
Related to #3108

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/functional_encrypted.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    def encrypt_(self, scheme):
+        """
+            This method will encrypt each value in the tensor using Functional encryption.
+
+        Args:
+            scheme: a PyFE Scheme created using
+                pyfe.scheme.Scheme()
+        """
+        input = self.child.flatten().tolist()
+        pk, msk = scheme.setup(vector_length=len(input))
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/functional_encrypted.py:43]: I guess the setup should not be depending of a Tensor. It should be done by a worker [kind of a crypto provider] which then shares the pk to the other workers for encryption. Andd it should never share the msk.

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 2]
```diff
@@ -1028,6 +1031,15 @@ def encrypt(self, protocol="mpc", **kwargs):
 
             return x_encrypted
 
+        elif protocol.lower() == "fe":
+            scheme = kwargs.get("scheme")
+
+            x = self.copy()
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/native.py:1037]: Do you think we have to copy?

File: syft/frameworks/torch/tensors/interpreters/functional_encrypted.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def encrypt(self, scheme):
+        """
+        This method will encrypt each value in the tensor using Functional encryption.
+
+        Args:
+            scheme: a PyFE Scheme created using
+                pyfe.scheme.Scheme()
+        """
+        output = FunctionalEncryptedTensor()
+        output.child = self.child
```
[the remark this answers is not in the document]
Comment (sukhadj) [reply syft/frameworks/torch/tensors/interpreters/functional_encrypted.py:31]: So we basically having <pyfe.vectors.EncryptedVector> as the child of the FET (like AST :laughing: ).
So it goes something like `(Wrapper)>FunctionalEncryptedTensor><pyfe.vectors.EncryptedVector object>` when we `.encypt()` on the torch Tensor. Do you think we should have should get rid of EncryptedVector for FET?

File: syft/frameworks/torch/tensors/interpreters/functional_encrypted.py (hunk 1 above)
Comment (sukhadj) [reply syft/frameworks/torch/tensors/interpreters/functional_encrypted.py:43]: It depends just on the one thing though - the length of the vector. But we should have a third party secure worker that generates the pk and msk and share only pk with worker. Will work on that :+1:

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (sukhadj) [reply syft/frameworks/torch/tensors/interpreters/native.py:1037]: Not sure. Tbh just followed PaillierTensor but I guess I will remove the copy

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -50,7 +50,18 @@ jobs:
...
+
+        sudo apt-get -y install flex bison libssl-dev python-dev libgmp-dev
+        wget http://crypto.stanford.edu/pbc/files/pbc-0.5.14.tar.gz
+        tar xf pbc-0.5.14.tar.gz
+        cd pbc-0.5.14
+        ./configure && make && sudo make install && cd ..
+        git clone https://github.com/JHUISI/charm.git
+        cd charm
+        ./configure.sh && make && sudo make install && sudo ldconfig && cd ..
+        sudo rm -rf charm/ pbc-0.5.14/ pbc-0.5.14.tar.gz
```
Comment (Syzygianinfern0) [inline .github/workflows/tests.yml:63]: @youben11

**Pull request closed without merging**